app/src/main/java/me/iacn/biliroaming/utils/UposReplaceHelper.kt

```diff
@@ -0,0 +1,226 @@
+package me.iacn.biliroaming.utils
+
+import android.net.Uri
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.future.future
+import me.iacn.biliroaming.BiliBiliPackage.Companion.instance
+import me.iacn.biliroaming.DashItemKt
+import me.iacn.biliroaming.R
+import me.iacn.biliroaming.StreamKt
+import me.iacn.biliroaming.UGCPlayViewReply
+import me.iacn.biliroaming.VideoInfoKt
+import me.iacn.biliroaming.XposedInit
+import me.iacn.biliroaming.copy
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+object UposReplaceHelper {
+    private val aliHost = string(R.string.ali_host)
+    private val cosHost = string(R.string.cos_host)
+    private val hwHost = string(R.string.hw_host)
+    private val aliovHost = string(R.string.aliov_host)
+    private val hwovHost = string(R.string.hwov_host)
+    private val hkBcacheHost = string(R.string.hk_bcache_host)
+
+    private val isLocatedCn by lazy {
+        (runCatchingOrNull { XposedInit.country.get(5L, TimeUnit.SECONDS) } ?: "cn") == "cn"
+    }
+
+    private val forceUpos = sPrefs.getBoolean("force_upos", false)
+    private val enablePcdnBlock = sPrefs.getBoolean("block_pcdn", false)
+
+    private lateinit var videoUposList: CompletableFuture<List<String>>
+    private val mainVideoUpos = sPrefs.getString("upos_host", null)?.let {
+        if (it == "\$1") null else it
+    } ?: if (isLocatedCn) hwHost else aliovHost
+    private val extraVideoUposList = if (isLocatedCn) {
+        when (mainVideoUpos) {
+            hwHost -> listOf(aliHost, cosHost)
+            aliHost -> listOf(hwHost, cosHost)
+            else -> listOf(aliHost, hwHost)
+        }
+    } else {
+        when (mainVideoUpos) {
+            aliovHost -> listOf(hkBcacheHost, hwovHost)
+            hkBcacheHost -> listOf(aliovHost, hwovHost)
+            else -> listOf(hkBcacheHost, aliovHost)
+        }
+    }
+    private val videoUposBase by lazy {
+        runCatchingOrNull { videoUposList.get(500L, TimeUnit.MILLISECONDS) }?.get(0)
+            ?: mainVideoUpos
+    }
+    private val videoUposBackups by lazy {
+        runCatchingOrNull { videoUposList.get(500L, TimeUnit.MILLISECONDS) }?.subList(1, 3)
+            ?: extraVideoUposList
+    }
+    private const val liveUpos = "c1--cn-gotcha01.bilivideo.com"
+
+    private val badVideoUposRegex by lazy {
+        Regex("""(szbdyd\.com)|(^(https?://)\w*\.mcdn\.bilivideo)""")
+    }
```

Comment:
这一个 contains 就行吧，有必要正则吗？
```suggestion
    private val String.isBadUpos = contains("szbdyd.com") || contains(".mcdn.bilivideo")
```